File: src/application/TransferCalculationService.ts
import { createCurrencyPair } from '../domain/CurrencyPair.js';
import { Money } from '../domain/Money.js';
import { TransferQuote } from '../domain/TransferQuote.js';
import { InvalidAmountError } from '../domain/errors/InvalidAmountError.js';
import { UnsupportedCurrencyPairError } from '../domain/errors/UnsupportedCurrencyPairError.js';
import { ExchangeRateProvider } from './ports/ExchangeRateProvider.js';
import { FeePolicy } from './ports/FeePolicy.js';

export interface TransferCalculationRequest {
  sourceCurrency: string;
  targetCurrency: string;
  amount: number;
}

export class TransferCalculationService {
  constructor(
    private readonly exchangeRateProvider: ExchangeRateProvider,
    private readonly feePolicy: FeePolicy
  ) {}

  async calculate(request: TransferCalculationRequest): Promise<TransferQuote> {
    if (request.amount <= 0) {
      throw new InvalidAmountError(request.amount);
    }

    const pair = createCurrencyPair(request.sourceCurrency, request.targetCurrency);
    const quote = await this.exchangeRateProvider.getQuote(pair);
    if (!quote) {
      throw new UnsupportedCurrencyPairError(pair.source, pair.target);
    }

    const amountToSend: Money = {
      amount: Number(request.amount.toFixed(2)),
      currency: pair.source
    };
    const amountToReceive: Money = {
      amount: Number((amountToSend.amount * quote.rate).toFixed(2)),
      currency: pair.target
    };
    const fee = this.feePolicy.calculateFee(amountToSend);
    const totalDebit: Money = {
      amount: Number((amountToSend.amount + fee.amount).toFixed(2)),
      currency: pair.source
    };

    return {
      pair,
      exchangeRate: quote.rate,
      amountToSend,
      amountToReceive,
      fee,
      totalDebit
    };
  }
}

File: internal/application/transfer_calculation_service.go
package application

import (
	"context"
	"errors"

	"github.com/shopspring/decimal"

	"github.com/example/payment-gateway/internal/application/ports"
	"github.com/example/payment-gateway/internal/domain"
)

// TransferCalculationInput defines the payload for computing a transfer quote.
type TransferCalculationInput struct {
	SourceCurrency string
	TargetCurrency string
	Amount         decimal.Decimal
}

// TransferCalculationService provides transfer quotes using the configured policies.
type TransferCalculationService struct {
	exchangeRateProvider ports.ExchangeRateProvider
	feePolicy            ports.FeePolicy
}

// NewTransferCalculationService creates a new service instance.
func NewTransferCalculationService(exchangeRateProvider ports.ExchangeRateProvider, feePolicy ports.FeePolicy) *TransferCalculationService {
	return &TransferCalculationService{
		exchangeRateProvider: exchangeRateProvider,
		feePolicy:            feePolicy,
	}
}

// Calculate produces a transfer quote from the supplied input.
func (s *TransferCalculationService) Calculate(ctx context.Context, input TransferCalculationInput) (domain.TransferQuote, error) {
	pair := domain.NewCurrencyPair(input.SourceCurrency, input.TargetCurrency)
	if !pair.IsValid() {
		return domain.TransferQuote{Allowed: false, Pair: pair}, errors.New("unsupported currency pair")
	}

	sendMoney, err := domain.NewMoney(pair.Base, input.Amount)
	if err != nil {
		return domain.TransferQuote{Allowed: false, Pair: pair}, err
	}

	rate, err := s.exchangeRateProvider.GetRate(ctx, pair)
	if err != nil {
		return domain.TransferQuote{Allowed: false, Pair: pair}, err
	}

	receiveAmount := domain.Money{Currency: pair.Quote, Amount: sendMoney.Amount.Mul(rate)}

	fee, err := s.feePolicy.CalculateFee(ctx, sendMoney)
	if err != nil {
		return domain.TransferQuote{Allowed: false, Pair: pair}, err
	}

	totalDebit, err := sendMoney.Add(fee)
	if err != nil {
		return domain.TransferQuote{Allowed: false, Pair: pair}, err
	}

	return domain.TransferQuote{
		Pair:          pair,
		ExchangeRate:  rate,
		SendAmount:    sendMoney,
		ReceiveAmount: receiveAmount,
		Fee:           fee,
		TotalDebit:    totalDebit,
		Allowed:       true,
	}, nil
}
